lib/translation.ts

```diff
@@ -0,0 +1,339 @@
+import {
+	getAppTranslations,
+	hasAppTranslations,
+	registerAppTranslations,
+	unregisterAppTranslations
+} from './registry'
+import type { Translations } from './registry'
+import { generateFilePath } from '@nextcloud/router'
+
+import DOMPurify from 'dompurify'
+import escapeHTML from 'escape-html'
+
+
+/**
+ * Returns the user's locale
+ */
+export function getLocale(): string {
+	return document.documentElement.dataset.locale || 'en'
+}
+
+export function getCanonicalLocale(): string {
+	return getLocale().replace(/_/g, '-')
+}
+
+/**
+ * Returns the user's language
+ */
+export function getLanguage(): string {
+	return document.documentElement.lang || 'en'
+}
+
+interface TranslationOptions {
+	/** enable/disable auto escape of placeholders (by default enabled) */
+	escape?: boolean,
+	/** enable/disable sanitization (by default enabled) */
+	sanitize?: boolean,
+}
+
+/**
+ * Translate a string
+ *
+ * @param {string} app the id of the app for which to translate the string
+ * @param {string} text the string to translate
+ * @param {object} vars map of placeholder key to value
+ * @param {number} number to replace %n with
+ * @param {object} [options] options object
+ * @return {string}
+ */
+export function translate(app: string, text: string, vars?: Record<string, any>, number?: number, options?: TranslationOptions): string {
+	const defaultOptions = {
+		escape: true,
+		sanitize: true,
+	}
+	const allOptions = Object.assign({}, defaultOptions, options || {})
+
+	const identity = (value) => value
+	const optSanitize = allOptions.sanitize ? DOMPurify.sanitize : identity
+	const optEscape = allOptions.escape ? escapeHTML : identity
+
+	// TODO: cache this function to avoid inline recreation
+	// of the same function over and over again in case
+	// translate() is used in a loop
+	const _build = (text: string, vars?: Record<string, any>, number?: number) => {
+		return text
+			.replace(/%n/g, '' + number)
+			.replace(/{([^{}]*)}/g, (match, key) => {
+				if (vars === undefined || !(key in vars)) return optSanitize(match)
```

Comment:
Please sue extended syntax for readability :)

```suggestion
				if (vars === undefined || !(key in vars)) {
					return optSanitize(match)
				}
```